python-checks/src/test/java/org/sonar/python/checks/MandatoryFunctionReturnTypeHintCheckTest.java

```diff
@@ -20,11 +20,170 @@
 import org.sonar.python.checks.utils.PythonCheckVerifier;
 
+import static org.sonar.python.checks.utils.CodeTestUtils.code;
+
 public class MandatoryFunctionReturnTypeHintCheckTest {
   @Test
   public void test() {
     PythonCheckVerifier.verify("src/test/resources/checks/mandatoryFunctionReturnType.py", new MandatoryFunctionReturnTypeHintCheck());
   }
+
+  @Test
+  public void testQuickFixAddReturnTypeLiteral() {
+    String codeWithIssue = code(
+      "def foo():",
+      "    return \"bar\"");
+    String codeFixed = code(
+      "def foo() -> str:",
+      "    return \"bar\"");
+    MandatoryFunctionReturnTypeHintCheck check = new MandatoryFunctionReturnTypeHintCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, MandatoryFunctionReturnTypeHintCheck.MESSAGE);
+  }
+
+  @Test
+  public void testQuickFixReturnExpression() {
+    String codeWithIssue = code(
+      "def foo(alice: str):",
+      "    return  alice + \"bar\"");
+    String codeFixed = code(
+      "def foo(alice: str) -> str:",
+      "    return  alice + \"bar\"");
+    MandatoryFunctionReturnTypeHintCheck check = new MandatoryFunctionReturnTypeHintCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, MandatoryFunctionReturnTypeHintCheck.MESSAGE);
+  }
+
+  @Test
+  public void testQuickFixNoReturnKeyword() {
+    String codeWithIssue = code(
+      "def foo(bob):",
+      "    print(bob) ");
+    String codeFixed = code(
+      "def foo(bob) -> None:",
+      "    print(bob) ");
+    MandatoryFunctionReturnTypeHintCheck check = new MandatoryFunctionReturnTypeHintCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, MandatoryFunctionReturnTypeHintCheck.MESSAGE);
+  }
+
+  @Test
+  public void testQuickFixNestedReturnType() {
+    String codeWithIssue = code(
+      "def foo(bob: bool):",
+      "    if bob:",
+      "        return 42",
+      "    else:",
+      "        return 40");
+    String codeFixed = code(
+      "def foo(bob: bool) -> int:",
+      "    if bob:",
+      "        return 42",
+      "    else:",
+      "        return 40");
+    MandatoryFunctionReturnTypeHintCheck check = new MandatoryFunctionReturnTypeHintCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, MandatoryFunctionReturnTypeHintCheck.MESSAGE);
+  }
+
+  @Test
+  public void testQuickFixGenericTypes() {
+    String codeWithIssue = code(
+      "def foo(alice: dict[str, int]):",
+      "    bob = {}",
+      "    for k, v in alice.items():",
+      "        bob[k] = str(v)",
+      "    return bob");
+    String codeFixed = code(
+      "def foo(alice: dict[str, int]) -> dict:",
```

Comment:
Not sure if this is the way to go. As Rule S6543 would then ask the user to provide a type parameter for `dict`. It feels a bit weird to quickly fix something which still needs fixing.

Reply:
Indeed, I agree. I think we can be conservative here and only offer the quick fix if the type is primitive (or at least non-generic) or None.

We could use a logic similar to generics without type argument check to see if the type is generic.

Reply:
Sadly, it is not as easy with the inferred type to detect if it could be a generic type. I think I will have to approach this by only taking care of nongeneric built-in types.